ntp-proto/src/packet.rs

```diff
@@ -113,25 +114,162 @@ pub struct NtpPacket<'a> {
+            Unknown { typeid, data } => Unknown {
                 typeid,
                 data: Cow::Owned(data.into_owned()),
             },
+            UniqueIdentifier(data) => UniqueIdentifier(Cow::Owned(data.into_owned())),
+            NtsCookie(data) => NtsCookie(Cow::Owned(data.into_owned())),
+            NtsCookiePlaceholder { body_length } => NtsCookiePlaceholder { body_length },
+            NtsEncryptedField { nonce, plaintext } => NtsEncryptedField {
+                nonce: Cow::Owned(nonce.into_owned()),
+                plaintext: Cow::Owned(plaintext.into_owned()),
+            },
         }
     }
 
-    fn serialize<W: std::io::Write>(&self, w: &mut W) -> std::io::Result<()> {
+    pub fn key_exchange_signature(nonce: &'a Nonce) -> Self {
+        // the real plaintext is the packet up to the start of this extension field
+        // it is inserted implicitly by the encoder
+        ExtensionField::NtsEncryptedField {
+            nonce: nonce.as_slice().into(),
+            plaintext: [].as_slice().into(),
+        }
+    }
+
+    pub fn serialize_without_encryption(&self, w: &mut Cursor<&mut [u8]>) -> std::io::Result<()> {
+        use aes_siv::{aead::KeyInit, Key};
+
+        let cipher = Aes256SivAead::new(Key::<Aes256SivAead>::from_slice([0; 64].as_slice()));
+
+        self.serialize(w, &cipher)
+    }
+
+    pub fn serialize(
+        &self,
+        w: &mut Cursor<&mut [u8]>,
+        cipher: &Aes256SivAead,
+    ) -> std::io::Result<()> {
+        let padding = [0; 4];
+
         match self {
+            ExtensionField::UniqueIdentifier(string) => {
+                w.write_all(&0x0104u16.to_be_bytes())?;
+                w.write_all(&(4 + string.len() as u16).to_be_bytes())?;
+                w.write_all(string)?;
+
+                let padding_bytes = next_multiple_of(string.len(), 4) - string.len();
+                w.write_all(&padding[..padding_bytes])?;
+            }
+            ExtensionField::NtsCookie(cookie) => {
+                w.write_all(&0x0204u16.to_be_bytes())?;
+                w.write_all(&(4 + cookie.len() as u16).to_be_bytes())?;
+                w.write_all(cookie)?;
+
+                let padding_bytes = next_multiple_of(cookie.len(), 4) - cookie.len();
+                w.write_all(&padding[..padding_bytes])?;
+            }
+            ExtensionField::NtsCookiePlaceholder { body_length } => {
+                w.write_all(&0x0304u16.to_be_bytes())?;
+                w.write_all(&(4 + body_length).to_be_bytes())?;
+
+                let zeros = [0u8; 100];
+                let mut remaining = next_multiple_of(*body_length as usize, 4);
+                while remaining > 0 {
+                    let n = usize::min(zeros.len(), remaining);
+                    w.write_all(&zeros[..n])?;
+                    remaining -= n;
+                }
+            }
+            ExtensionField::NtsEncryptedField {
+                nonce,
+                plaintext: _,
```

Comment:
this is kind of weird. The plaintext is implicit when encoding, but I suspect you'll want to be able to access it when decoding (otherwise what is even the point).